tests/scanner/audit/external_project_access_engine_test.py

```diff
@@ -0,0 +1,220 @@
+
+
+class ExternalProjectAccessRuleBookTest(ForsetiTestCase):
+    """Tests for the ExternalProjectAccessRuleBook."""
+
+    TEST_GOOD_RULE = dict(name='default', allowed_ancestors=['organizations/7890'])
+    TEST_BAD_RULE = dict(name='default', allowed_ancestors=['policy/12345'])
+    TEST_RULE_DEFS = dict(rules=[TEST_GOOD_RULE])
+
+    TEST_ANCESTORS = [Project("123"), Folder("456"), Organization("7890")]
+    TEST_BAD_ANC = [Project("123"), Folder("456"), Organization("ABC")]
+
+    def setUp(self):
+        """Set up."""
+        self.rule_index = 0
+        self.epare = epare
+        self.epare.LOGGER = mock.MagicMock()
+        self.inventory_config = mock.MagicMock()
+        
+        self.inventory_config.get_root_resource_id = mock.MagicMock(return_value='organizations/7890')
+        self.rule_book = epare.ExternalProjectAccessRuleBook(self.inventory_config)
+
+    def test_validate_good_ancestor(self):
+        try:
+            self.rule_book.validate_ancestors(ExternalProjectAccessRuleBookTest.TEST_GOOD_RULE['allowed_ancestors'], 0)
+        except:
+            self.fail("Unexpected exception thrown")
+    
+    def test_validate_bad_ancestor(self):
+        with self.assertRaises(audit_errors.InvalidRulesSchemaError): 
+            self.rule_book.validate_ancestors(ExternalProjectAccessRuleBookTest.TEST_BAD_RULE['allowed_ancestors'], 0)
+
+    def test_process_good_rule(self):
+        try:
+            resources = self.rule_book.process_rule(ExternalProjectAccessRuleBookTest.TEST_GOOD_RULE, 0)
+            self.assertEqual(resources[0].id, '7890')
+            self.assertTrue(isinstance(resources[0], Organization))
+        except Exception as e:
+            self.fail("Unexpected exception thrown: " + str(e))
+
+    def test_process_bad_rule(self):
+        with self.assertRaises(audit_errors.InvalidRulesSchemaError): 
+            self.rule_book.process_rule(ExternalProjectAccessRuleBookTest.TEST_BAD_RULE, 0)
+
+    def test_add_rule(self):
+        self.rule_book.add_rule(ExternalProjectAccessRuleBookTest.TEST_GOOD_RULE, 0)
+        self.assertEqual(1, len(self.rule_book.resource_rules_map))
+
+    def test_add_rules(self):
+        self.rule_book.add_rules(self.TEST_RULE_DEFS)
+        self.assertEqual(1, len(self.rule_book.resource_rules_map))
+
+    def test_no_viloations(self):
+
+        violations = self.rule_book.find_policy_violations("user@example.com", self.TEST_ANCESTORS)
+
+        self.assertEqual(0, len(violations))
+
+    def test_viloations(self):
+
+        violations = self.rule_book.find_policy_violations("user@example.com", self.TEST_BAD_ANC)
+
+        self.assertEqual(0, len(violations))
+
+class ExternalProjectAccessRuleTest(ForsetiTestCase):
+    """Tests for the ExternalProjectAccessRuleBook."""
+
+    TEST_ANCESTORS = [Project("123"), Folder("456"), Organization("7890")]
+        
+    def test_single_item_in_rule_match(self):
+        rule = epare.Rule(rule_name="test_single_item_in_rule_match",
```

Comment:
You should use a `self.epare` or `self.rule_engine`.

Reply:
with rules_engine being a module, is that appropriate?